### PR TITLE
Add result guard to waypoints_to_commands() and add_takeoff_and_landi…

### PR DESCRIPTION
**Pull request by applepie7864**

### CODE REVIEW

File: modules/add_takeoff_and_landing_command.py [hunk 1]
```diff
@@ -25,9 +25,13 @@ def add_takeoff_and_landing_command(commands: "list[dronekit.Command]",
 
     Returns
     -------
-    list[dronekit.Command]
-        Dronekit commands with takeoff and land commands that can be sent to the drone.
+    tuple[bool, list[dronekit.Command] | None]: 
+        (False, None) if empty commands list,
+        (True, dronekit commands with takeoff and land commands that can be sent to the drone) otherwise
     """
+    if not commands:
```
Comment (Xierumeng) [inline modules/add_takeoff_and_landing_command.py:32]: Prefer `if len(commands) == 0:`

File: modules/waypoints_to_commands.py [hunk 2]
```diff
@@ -24,9 +24,14 @@ def waypoints_to_commands(waypoints: "list[tuple[float, float]]",
 
     Returns
     -------
-    list[dronekit.Command]
+    tuple[bool, list[dronekit.Command] | None]: 
+        (False, None) if empty waypoints list,
+        (True, dronekit commands that can be sent to the drone) otherwise
         dronekit commands that can be sent to the drone.
     """
+    if not waypoints:
```
Comment (Xierumeng) [inline modules/waypoints_to_commands.py:32]: Prefer `if len(waypoints) == 0:`

**Review by Xierumeng — COMMENTED** [review]
Good start! Do all the tests pass? Would `path_2023.py` run with these changes?

File: modules/waypoints_to_commands.py (hunk 2 above)
Comment (annie-eonmedia) [reply modules/waypoints_to_commands.py]: Hi, I've edited both

**Comment by applepie7864** [comment]
Hi Xierumeng,
I've edited both of the above. I can't run the tests or path_2023.py since I have MacOS system and one of the tools needed to run the program is Mission Planner that is only available for Windows, Android and Linux. Is there a way to work around this? Thanks.

File: path_2023.py [hunk 3]
```diff
@@ -43,12 +43,12 @@ def run() -> int:
         print("ERROR: waypoint_names_to_coordinates")
         return -1
     
-    waypoint_commands = waypoints_to_commands.waypoints_to_commands(waypoints, ALTITUDE)
+    result, waypoint_commands = waypoints_to_commands.waypoints_to_commands(waypoints, ALTITUDE)
     if len(waypoint_commands) == 0:
```
Comment (Xierumeng) [inline path_2023.py:47]: Change this to check `result` .

File: path_2023.py [hunk 4]
```diff
@@ -43,12 +43,12 @@ def run() -> int:
...
         print("Error: waypoints_to_commands")
         return -1
     
-    takeoff_landing_commands = add_takeoff_and_landing_command.add_takeoff_and_landing_command(waypoint_commands, ALTITUDE)
+    result, takeoff_landing_commands = add_takeoff_and_landing_command.add_takeoff_and_landing_command(waypoint_commands, ALTITUDE)
     if len(takeoff_landing_commands) == 0:
```
Comment (Xierumeng) [inline path_2023.py:52]: Change this to check `result` .

File: tests/test_add_takeoff_and_landing_command.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 import dronekit
 import pytest
 
-from modules import add_takeoff_and_landing_command
+import modules.add_takeoff_and_landing_command as add_takeoff_and_landing_command
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:9]: Change this back.

File: tests/test_add_takeoff_and_landing_command.py [hunk 6]
```diff
@@ -76,39 +76,47 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
     """
     Helper function to assert the correctness of takeoff and landing commands.
     """
-    assert isinstance(commands_actual, list)
+    if isinstance(commands_actual, type(None)):
+        assert len(commands_expected) == 0
+    else:
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:81]: Remove this.

File: tests/test_add_takeoff_and_landing_command.py [hunk 7]
```diff
@@ -76,39 +76,47 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
+        # Test original commands
+        assert len(commands_actual) == len(commands_expected) + 2
+        assert commands_actual[1:-1] == commands_expected
 
 
 def test_add_takeoff_and_landing_on_empty_commands(empty_commands: "list[dronekit.Command]"):
     """
     Tests functionality correctness of add_takeoff_and_landing_command on empty list of commands.
     """
     commands_expected = copy.deepcopy(empty_commands)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:106]: Remove this.

File: tests/test_add_takeoff_and_landing_command.py [hunk 8]
```diff
@@ -76,39 +76,47 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
-    commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
+    result, commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
+    
+    assert(not result)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:109]: Remove unnecessary parantheses.

File: tests/test_add_takeoff_and_landing_command.py [hunk 9]
```diff
@@ -76,39 +76,47 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
     assert_expected_takeoff_and_landing_commands(commands_actual, commands_expected, ALTITUDE)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:110]: `assert actual is None`

File: tests/test_add_takeoff_and_landing_command.py [hunk 10]
```diff
@@ -76,39 +76,47 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
 
+test_add_takeoff_and_landing_on_empty_commands([])
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:112]: Remove this.

File: tests/test_add_takeoff_and_landing_command.py [hunk 11]
```diff
@@ -76,39 +76,47 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
 
 def test_add_takeoff_and_landing_on_nonempty_commands(non_empty_commands: "list[dronekit.Command]"):
     """
     Tests functionality correctness of add_takeoff_and_landing_command on non-empty list of commands.
     """
     commands_expected = copy.deepcopy(non_empty_commands)
-    commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(non_empty_commands, ALTITUDE)
+    result, commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(non_empty_commands, ALTITUDE)
+    
+    assert(result)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:121]: Remove unnecessary parantheses.

File: tests/test_waypoints_to_commands.py [hunk 12]
```diff
@@ -4,7 +4,7 @@
 
 import dronekit
 
-from modules import waypoints_to_commands
+import modules.waypoints_to_commands as waypoints_to_commands
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:7]: Change this back.

File: tests/test_waypoints_to_commands.py
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:1]: Add a test for empty waypoints list.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: tests/test_add_takeoff_and_landing_command.py [hunk 13]
```diff
@@ -6,7 +6,7 @@
 import dronekit
 import pytest
 
-from modules import add_takeoff_and_landing_command
+import modules as add_takeoff_and_landing_command
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:9]: `from modules import ...`

File: tests/test_add_takeoff_and_landing_command.py [hunk 14]
```diff
@@ -91,24 +90,34 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
     assert landing_command.frame == add_takeoff_and_landing_command.MAVLINK_LANDING_FRAME
     assert landing_command.command == add_takeoff_and_landing_command.MAVLINK_LANDING_COMMAND
 
-    # Test original commands
+     # Test original commands
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:93]: Remove 1 space before `#` .

File: tests/test_add_takeoff_and_landing_command.py [hunk 15]
```diff
@@ -91,24 +90,34 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
     assert len(commands_actual) == len(commands_expected) + 2
     assert commands_actual[1:-1] == commands_expected
 
+def assert_expected_takeoff_and_landing_commands_empty(commands_actual: "list[dronekit.Command]",
+                                                       commands_expected: "list[dronekit.Command]"):
+    """
+    Helper function to assert the correctness of takeoff and landing commands when there are no commands
+    """
+    assert(commands_actual is None)
+    assert(len(commands_expected) == 0)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:103]: This is unnecessary.

File: tests/test_add_takeoff_and_landing_command.py [hunk 16]
```diff
@@ -91,24 +90,34 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
 
 def test_add_takeoff_and_landing_on_empty_commands(empty_commands: "list[dronekit.Command]"):
     """
     Tests functionality correctness of add_takeoff_and_landing_command on empty list of commands.
     """
     commands_expected = copy.deepcopy(empty_commands)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:109]: Remove this.

File: tests/test_add_takeoff_and_landing_command.py [hunk 17]
```diff
@@ -91,24 +90,34 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
-    commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
-    assert_expected_takeoff_and_landing_commands(commands_actual, commands_expected, ALTITUDE)
+    result, commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
 
+    assert(not result)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:112]: Remove unnecessary parantheses.

File: tests/test_add_takeoff_and_landing_command.py [hunk 18]
```diff
@@ -91,24 +90,34 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
+    assert_expected_takeoff_and_landing_commands_empty(commands_actual, commands_expected)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:113]: `assert actual is None`

File: tests/test_waypoints_to_commands.py [hunk 19]
```diff
@@ -4,7 +4,7 @@
 
 import dronekit
 
-from modules import waypoints_to_commands
+import modules as waypoints_to_commands
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:7]: `from modules import ...`

File: tests/test_waypoints_to_commands.py [hunk 20]
```diff
@@ -4,7 +4,7 @@
...
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:9]: Add a test for empty waypoints list.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: tests/test_add_takeoff_and_landing_command.py [hunk 21]
```diff
@@ -76,8 +76,7 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
     """
     Helper function to assert the correctness of takeoff and landing commands.
     """
-    assert isinstance(commands_actual, list)
-
+
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:79]: Restore `assert_expected_takeoff_and_landing_commands()` to its original state.

File: tests/test_add_takeoff_and_landing_command.py [hunk 22]
```diff
@@ -95,20 +94,22 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
     assert len(commands_actual) == len(commands_expected) + 2
     assert commands_actual[1:-1] == commands_expected
 
-
 def test_add_takeoff_and_landing_on_empty_commands(empty_commands: "list[dronekit.Command]"):
     """
     Tests functionality correctness of add_takeoff_and_landing_command on empty list of commands.
     """
     commands_expected = copy.deepcopy(empty_commands)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:101]: Remove line 103.

File: tests/test_add_takeoff_and_landing_command.py [hunk 23]
```diff
@@ -95,20 +94,22 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
-    commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
-    assert_expected_takeoff_and_landing_commands(commands_actual, commands_expected, ALTITUDE)
+    result, commands_actual = add_takeoff_and_landing_command.add_takeoff_and_landing_command(empty_commands, ALTITUDE)
 
+    assert(commands_actual is None)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:104]: Add assertion that result is false.

File: tests/test_add_takeoff_and_landing_command.py [hunk 24]
```diff
@@ -95,20 +94,22 @@ def assert_expected_takeoff_and_landing_commands(commands_actual: "list[dronekit
...
+    assert(len(commands_expected) == 0)
```
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:105]: Remove unnecessary parantheses.

Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:105]: Remove this line.

File: tests/test_waypoints_to_commands.py
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:1]: Add another test for empty waypoints list.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: tests/test_add_takeoff_and_landing_command.py (hunk 22 above)
Comment (Xierumeng) [reply tests/test_add_takeoff_and_landing_command.py:102]: Apologies, I misread. Should be line 102.

File: tests/test_waypoints_to_commands.py
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:1]: Add periods to each docstring (e.g. `Test Process.` , `Tests functionality ...` ).

File: tests/test_waypoints_to_commands.py [hunk 25]
```diff
@@ -6,6 +6,17 @@
 
 from modules import waypoints_to_commands
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:8]: 2 empty lines between imports and top level functions.

File: tests/test_waypoints_to_commands.py [hunk 26]
```diff
@@ -6,6 +6,17 @@
...
+    """
+    Tests functionality correctness of waypoints_to_commands on empty input
+    """
+    waypoints = []
+    altitude = 100
+    
+    result, commands_actual = waypoints_to_commands.waypoints_to_commands(waypoints, altitude)
+    
+    assert not result
+    assert commands_actual is None
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:20]: 2 empty lines between top level functions.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: tests/test_waypoints_to_commands.py [hunk 27]
```diff
@@ -1,20 +1,29 @@
 """
-Test process
+Test process.
 """
-
 import dronekit
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:4]: Add 1 empty line between docstring and package imports.

File: tests/test_waypoints_to_commands.py (hunk 25 above)
Comment (Xierumeng) [reply tests/test_waypoints_to_commands.py:6]: Add a 2nd empty line between imports and top level functions.

File: tests/test_waypoints_to_commands.py [hunk 28]
```diff
@@ -1,20 +1,29 @@
...
-
 from modules import waypoints_to_commands
 
+def test_waypoints_to_commands_empty_input():
+    """
+    Tests functionality correctness of waypoints_to_commands on empty input.
+    """
+    waypoints = []
+    altitude = 100
+    result, commands_actual = waypoints_to_commands.waypoints_to_commands(waypoints, altitude)
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:13]: Add empty line above this line.

File: tests/test_waypoints_to_commands.py [hunk 29]
```diff
@@ -1,20 +1,29 @@
...
+    assert not result
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:14]: Add empty line above this line.

File: tests/test_waypoints_to_commands.py (hunk 26 above)
Comment (Xierumeng) [reply tests/test_waypoints_to_commands.py:16]: Add a 2nd empty line between end of top level function and beginning of top level function.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: modules/add_takeoff_and_landing_command.py [hunk 30]
```diff
@@ -12,7 +12,7 @@
 
 
 def add_takeoff_and_landing_command(commands: "list[dronekit.Command]",
-                                    altitude: float) -> "list[dronekit.Command]":
+                                    altitude: float) -> "tuple[bool, list[dronekit.Command] | None]":
     """
     Prepends a takeoff command and appends a landing command to a list of dronekit commands
```
Comment (Xierumeng) [inline modules/add_takeoff_and_landing_command.py:17]: Add period.

File: modules/add_takeoff_and_landing_command.py [hunk 31]
```diff
@@ -25,9 +25,13 @@ def add_takeoff_and_landing_command(commands: "list[dronekit.Command]",
 
     Returns
     -------
-    list[dronekit.Command]
-        Dronekit commands with takeoff and land commands that can be sent to the drone.
+    tuple[bool, list[dronekit.Command] | None]: 
+        (False, None) if empty commands list,
+        (True, dronekit commands with takeoff and land commands that can be sent to the drone) otherwise
```
Comment (Xierumeng) [inline modules/add_takeoff_and_landing_command.py:30]: Add period.

File: modules/waypoints_to_commands.py
Comment (Xierumeng) [inline modules/waypoints_to_commands.py:1]: Add period to line 2.

File: modules/waypoints_to_commands.py [hunk 32]
```diff
@@ -24,9 +24,14 @@ def waypoints_to_commands(waypoints: "list[tuple[float, float]]",
 
     Returns
     -------
-    list[dronekit.Command]
+    tuple[bool, list[dronekit.Command] | None]: 
+        (False, None) if empty waypoints list,
+        (True, dronekit commands that can be sent to the drone) otherwise
```
Comment (Xierumeng) [inline modules/waypoints_to_commands.py:29]: Add period.

File: modules/waypoints_to_commands.py [hunk 33]
```diff
@@ -24,9 +24,14 @@ def waypoints_to_commands(waypoints: "list[tuple[float, float]]",
...
         dronekit commands that can be sent to the drone.
```
Comment (Xierumeng) [inline modules/waypoints_to_commands.py:30]: Remove this line.

File: tests/test_add_takeoff_and_landing_command.py
Comment (Xierumeng) [inline tests/test_add_takeoff_and_landing_command.py:1]: Add period to line 2.

File: tests/test_waypoints_to_commands.py [hunk 34]
```diff
@@ -1,20 +1,34 @@
 """
-Test process
+Test process.
 """
 
 import dronekit
 
 from modules import waypoints_to_commands
 
+def test_waypoints_to_commands_empty_input():
```
Comment (Xierumeng) [inline tests/test_waypoints_to_commands.py:9]: Add a 2nd empty line above this line.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

**Review by Xierumeng — DISMISSED** [review]
~~Approved~~ Will approve once I test.

**Review by Xierumeng — APPROVED** [review]
Approved, tested working.
